algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs

```diff
@@ -69,32 +77,43 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         rng: &mut R,
     ) -> Result<prover::State<'a, F, MM>, AHPError> {
         let round_time = start_timer!(|| "AHP::Prover::FirstRound");
-        let constraint_domain = state.constraint_domain;
-        let batch_size = state.batch_size;
-
-        let z_a = state.z_a.take().unwrap();
-        let z_b = state.z_b.take().unwrap();
-        let private_variables = core::mem::take(&mut state.private_variables);
-        assert_eq!(z_a.len(), batch_size);
-        assert_eq!(z_b.len(), batch_size);
-        assert_eq!(private_variables.len(), batch_size);
-        let mut r_b_s = Vec::with_capacity(batch_size);
-
-        let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(3 * batch_size);
-        let state_ref = &state;
-        for (i, (z_a, z_b, private_variables, x_poly)) in
-            itertools::izip!(z_a, z_b, private_variables, &state.x_poly).enumerate()
-        {
-            job_pool.add_job(move || Self::calculate_w(witness_label("w", i), private_variables, x_poly, state_ref));
-            job_pool.add_job(move || Self::calculate_z_m(witness_label("z_a", i), z_a, false, state_ref, None));
-            let r_b = F::rand(rng);
-            job_pool.add_job(move || Self::calculate_z_m(witness_label("z_b", i), z_b, true, state_ref, Some(r_b)));
-            if MM::ZK {
-                r_b_s.push(r_b);
+        let mut r_b_s = Vec::with_capacity(state.circuit_specific_states.len());
+        let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(3 * state.total_instances);
+        for (circuit, circuit_state) in state.circuit_specific_states.iter_mut() {
+            let batch_size = circuit_state.batch_size;
+
+            let z_a = circuit_state.z_a.take().unwrap();
+            let z_b = circuit_state.z_b.take().unwrap();
+            let private_variables = core::mem::take(&mut circuit_state.private_variables);
+            let x_polys = circuit_state.x_polys.clone();
+            assert_eq!(z_a.len(), batch_size);
+            assert_eq!(z_b.len(), batch_size);
+            assert_eq!(private_variables.len(), batch_size);
+            assert_eq!(x_polys.len(), batch_size);
+
+            let c_domain = circuit_state.constraint_domain;
+            let i_domain = circuit_state.input_domain;
+
+            let mut circuit_r_b_s = Vec::with_capacity(batch_size);
+            for (j, (z_a, z_b, private_variables, x_poly)) in
+                itertools::izip!(z_a, z_b, private_variables, x_polys).enumerate()
+            {
+                let w_label = witness_label(circuit.id, "w", j);
+                let za_label = witness_label(circuit.id, "z_a", j);
+                let zb_label = witness_label(circuit.id, "z_b", j);
+                job_pool.add_job(move || {
+                    Self::calculate_w(w_label, private_variables, &x_poly, c_domain, i_domain, circuit)
+                });
```

Comment:
```suggestion
            for (j, (z_a, z_b, private_vars, x_poly)) in
                itertools::izip!(z_a, z_b, private_variables, x_polys).enumerate()
            {
                let w_label = witness_label(circuit.id, "w", j);
                let za_label = witness_label(circuit.id, "z_a", j);
                let zb_label = witness_label(circuit.id, "z_b", j);
                job_pool.add_job(move || Self::calculate_w(w_label, private_vars, &x_poly, c_domain, i_domain, circuit));
```

Reply:
The line is too long for `cargo fmt`'s taste

Reply:
ACK shortening some variable/function names